### PR TITLE
fix: resolve OS-default device before showing mic warning (#50)

### LINKED ISSUE #50: Mic warning fires when using OS-default device, even if that default is a proper audio interface
## Summary

The app shows a 'low-quality mic' style warning whenever the selected input is the macOS *default* device — even when that default is actually pointing at a proper audio interface. The warning is meant to nudge users away from built-in mics, but it currently can't tell apart 'default = built-in MacBook mic' (warn) from 'default = my Focusrite' (don't warn).

Result: false-positive warnings that scare users away from their best available input.

## Repro
1. Connect external audio interface on macOS
2. Set it as system default (System Settings → Sound → Input)
3. Join a session in cozytrack
4. **Expected:** no warning, you're already on the best input
5. **Actual:** warning displayed implying the input is poor

## Why it happens

Browsers expose a special audio device with `deviceId === 'default'` whose label is something like `'Default - <real device name>'`. The 'real' device is *also* enumerated separately with its own deviceId. So when the OS default is in use, the warning logic likely:

- Triggers on `deviceId === 'default'` regardless of what hardware that resolves to, OR
- Substring-matches 'default' / 'built-in' / 'macbook' in the label, which also fires for `'Default - <good device>'`

Either way the underlying hardware isn't being checked.

## Fix direction

Resolve the OS-default to its underlying device before deciding whether to warn:

1. Call `navigator.mediaDevices.enumerateDevices()` and find the entry with `deviceId === 'default'`
2. Parse the label after `'Default - '` to get the underlying device name, OR match its `groupId` against other enumerated entries to find the corresponding real device
3. Apply the warning rule to that **resolved** device, not to the literal 'default' entry

A known-bad list (built-in MacBook mic, AirPods, generic bluetooth headsets) is reasonable as a starting point. The warning should only fire when the resolved underlying device is on that list.

### Even better (longer-term)
Don't rely on device names at all. Add a 'mic check' flow that:
- Records 3 seconds of silence + 3 seconds of speech
- Measures noise floor, dynamic range, sample rate, channel count
- Warns based on actual signal quality, not vendor strings

That avoids both false positives (good interface mistaken for built-in) and false negatives (cheap USB mic with a brand name we don't recognize).

## Suspect file
`src/app/studio/[id]/page.tsx` or wherever device selection / warning UI lives. Look for substring checks against 'default' or 'built-in' in track labels.

## Related
- Discovered during the 2026-04-27 cohosted recording test session

### PR DESCRIPTION
**Pull request by pashafateev**
## Summary

The built-in mic warning fires whenever the selected input is the browser's virtual `default` device — even when that default points at a real audio interface (Focusrite, Shure, etc.). Root cause: `isBuiltInMic` substring-matched the literal string `"default"` in device labels, so `"Default - Focusrite Scarlett 2i2"` was treated identically to `"MacBook Pro Microphone"`.

## Fix

- Add `resolveDefaultDevice(devices, deviceId)` in `src/lib/devices.ts`. When the selection is the `deviceId === "default"` entry, look up the underlying real device by `groupId`, falling back to parsing `"Default - <name>"` from the label.
- Add `isSelectedMicBuiltIn(devices, deviceId)` that resolves first, then applies the built-in keyword check.
- Drop `"default"` from `BUILTIN_KEYWORDS` so the literal "Default" label can no longer trigger the warning on its own.
- Update `src/app/studio/[id]/page.tsx` to call `isSelectedMicBuiltIn(mics, selectedMic)` at both the join-time check and the `selectedMicIsBuiltIn` prop.

The longer-term signal-quality mic check described in the issue is intentionally out of scope here — this PR just kills the false positive.

## Testing

- New `tests/devices.test.ts` (11 cases) covering:
  - `default → real interface` resolves correctly via `groupId` and via label parsing
  - selecting the OS default with a Focusrite as underlying device does **not** warn
  - selecting the OS default with the MacBook mic as underlying device **does** warn
  - explicitly selecting the built-in mic still warns
  - non-default selections, missing devices, edge cases
- `npm test -- tests/devices.test.ts` → 11/11 pass
- `npm run lint` → clean
- `npx tsc --noEmit` → clean

Manual macOS repro is recommended before merging: connect external interface, set as system default, join a session, confirm no warning. The logic is covered by the unit tests but the OS/browser interaction is what the original bug is about.

Closes #50

### CODE REVIEW

File: src/lib/devices.ts [hunk 1]
```diff
@@ -2,9 +2,62 @@
...
+  // with that name (Chrome wraps with "Default - ", Safari/Firefox vary).
+  const match = selected.label.match(/^Default\s*[-–—]\s*(.+)$/i);
+  if (match) {
+    const underlyingName = match[1].trim().toLowerCase();
+    const byLabel = devices.find(
+      (d) =>
+        d.deviceId !== "default" &&
+        d.label.toLowerCase().includes(underlyingName),
+    );
+    if (byLabel) return byLabel;
```
[the remark this answers is not in the document]
Comment (pashafateev) [reply src/lib/devices.ts]: Added a guard so empty/whitespace-only `underlyingName` skips the label fallback and the default entry is returned unchanged. Fixed in 6885af8.

File: src/lib/devices.ts [hunk 2]
```diff
@@ -2,9 +2,62 @@
...
+  if (selected.deviceId !== "default") return selected;
+
+  // Try matching by groupId — the most reliable signal.
+  const byGroup = devices.find(
+    (d) => d.deviceId !== "default" && d.groupId && d.groupId === selected.groupId,
+  );
+  if (byGroup) return byGroup;
+
+  // Fallback: parse "Default - <name>" and match a device whose label ends
+  // with that name (Chrome wraps with "Default - ", Safari/Firefox vary).
```
[the remark this answers is not in the document]
Comment (pashafateev) [reply src/lib/devices.ts:36]: Switched the matcher from `includes` to `endsWith` so the implementation matches the comment (the underlying name is at the end of `Default - <name>`). Fixed in 6885af8.

File: tests/devices.test.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+  it("falls back to label parsing when groupIds don't match", () => {
+    const devices = [
+      dev("default", "Default - Focusrite Scarlett 2i2", ""),
+      dev("abc123", "Focusrite Scarlett 2i2", ""),
+    ];
+    const resolved = resolveDefaultDevice(devices, "default");
+    expect(resolved?.deviceId).toBe("abc123");
+  });
+
```
[the remark this answers is not in the document]
Comment (pashafateev) [reply tests/devices.test.ts:61]: Added the suggested unit test covering the `Default - ` (empty underlying name) case alongside other devices — expects the default entry to be returned unchanged. Fixed in 6885af8.
